backend/inspirehep/records/api/literature.py

```diff
@@ -267,106 +269,161 @@ def add_files(self, data):
-                )
-        return builder.record.get("figures")
+        return tasks
 
-    def add_file(
-        self, url, original_url=None, key=None, filename=None, *args, **kwargs
-    ):
-        if current_s3_instance.is_s3_url(url) and not current_app.config.get(
-            "UPDATE_S3_FILES_METADATA", False
-        ):
-            result = {}
-            if key not in url:
-                filename = filename or key
-                key = url.split("/")[-1]
-                result.update({"key": key, "filename": filename})
-            LOGGER.info(
-                "File already on S3 - Skipping",
-                url=url,
-                key=key,
-                recid=self.get("control_number"),
-                uuid=self.id,
-            )
-            return result
-        file_data = download_file_from_url(url)
-        new_key = hash_data(file_data)
-        mimetype = magic.from_buffer(file_data, mime=True)
-        file_data = BytesIO(file_data)
-        filename = filename or key
-        acl = current_app.config["S3_FILE_ACL"]
-        if current_s3_instance.file_exists(new_key):
-            LOGGER.info(
-                "Replacing file metadata",
-                key=new_key,
+    def _update_file_entry(self, file_data, add_file_func):
+        try:
+            add_file_func(**file_data)
+        except ValueError:
+            LOGGER.exception(
+                "Duplicated file found",
                 recid=self.get("control_number"),
                 uuid=self.id,
+                file=file_data.get("key"),
             )
-            current_s3_instance.replace_file_metadata(new_key, filename, mimetype, acl)
-        else:
-            LOGGER.info(
-                "Uploading file to s3",
-                key=new_key,
+
+    def _process_tasks_results(self, tasks, files, add_file_func):
+        try:
+            for future in as_completed(
+                tasks, timeout=current_app.config.get("FILES_UPLOAD_THREAD_TIMEOUT", 30)
+            ):
+                idx = tasks[future]
+                new_file_data = future.result()
+                files[idx].update(new_file_data)
+                self._update_file_entry(files[idx], add_file_func)
+        except TimeoutError:
+            LOGGER.exception(
+                "Threads didn't return results on time",
                 recid=self.get("control_number"),
                 uuid=self.id,
             )
-            current_s3_instance.upload_file(file_data, new_key, filename, mimetype, acl)
-        result = {
-            "key": new_key,
-            "filename": filename,
-            "url": current_s3_instance.get_file_url(new_key),
-        }
-        if (
-            url.startswith("http")
-            and not current_s3_instance.is_s3_url(url)
-            and not original_url
-        ):
-            result["original_url"] = url
-        return result
+            raise
+
+    def add_files_data(self, files, builder_add_file):
```

Comment:
This can be moved to ``add_files`` and have only one threadpool for both documents and figures

Reply:
One thread pool for both figures and documents is not good idea IMHO as results from them are going to separate keys in the record. I could put them in one but then instead of having two simple loops for putting results in record there will be one more complex loop with if/else and results themselves would be put in more complex structure to hold information where results should be put.